core/trino-main/src/test/java/io/trino/server/security/TestResourceSecurity.java

```diff
@@ -190,13 +202,66 @@ public void testPasswordAuthenticator()
                         .build())
                 .build()) {
             server.getInstance(Key.get(PasswordAuthenticatorManager.class)).setAuthenticator(TestResourceSecurity::authenticate);
-            server.getInstance(Key.get(AccessControlManager.class)).addSystemAccessControl(new TestSystemAccessControl());
+            server.getInstance(Key.get(AccessControlManager.class)).addSystemAccessControl(new TestSystemAccessControl(false));
             HttpServerInfo httpServerInfo = server.getInstance(Key.get(HttpServerInfo.class));
             assertAuthenticationDisabled(httpServerInfo.getHttpUri());
             assertPasswordAuthentication(httpServerInfo.getHttpsUri());
         }
     }
 
+    @Test
+    public void testPasswordAuthenticatorUserMapping()
+            throws Exception
+    {
+        try (TestingTrinoServer server = TestingTrinoServer.builder()
+                .setProperties(ImmutableMap.<String, String>builder()
+                        .putAll(SECURE_PROPERTIES)
+                        .put("http-server.authentication.type", "password")
+                        .put("http-server.authentication.password.user-mapping.pattern", ALLOWED_USER_MAPPING_PATTERN)
+                        .build())
+                .setAdditionalModule(binder -> jaxrsBinder(binder).bind(TestResource.class))
+                .build()) {
+            server.getInstance(Key.get(PasswordAuthenticatorManager.class)).setAuthenticator(TestResourceSecurity::authenticate);
+            server.getInstance(Key.get(AccessControlManager.class)).addSystemAccessControl(new TestSystemAccessControl(false));
+            HttpServerInfo httpServerInfo = server.getInstance(Key.get(HttpServerInfo.class));
+
+            // Test sets basic auth user and X-Trino-User, and the authenticator is performing user mapping.
+            // Normally this would result in an impersonation check to the X-Trino-User, but the password
+            // authenticator has a hack to clear X-Trino-User in this case.
```

Comment:
I don't think we should skip allowed mapping in this case. User can just `|(.*)` or something to their mapping.

let's not loosen the restrictions if not necessary

Reply:
The problem this PR is solving, which needs to be explained in the commit message, is that user mapping for password authentication is useless, since clients always send the `X-Trino-User` header. For example, suppose that we use the CLI to connect with user `foo`, and the administrator has configured user `foo` to map to `bar`. This is what happens today:

Run CLI: `trino --user foo --password --execute 'SELECT ...'`

Query request (skipping base64 for the `Authorization` header for readability):
```
Authorization: Basic foo:mypassword
X-Trino-User: foo
```

This happens on the server:

* user `foo` is authenticated using the `Authorization` header
* user `foo` is mapped to user `bar`
* user `bar` is mapped back to user `foo` using the `X-Trino-User` header
  * this first does an impersonation check to see if user `bar` can become user `foo`
  * we then set the user back to `foo`

Without changing all of the clients, there is no way for user mapping to work. We can safely assume that impersonating the password user was not intended behavior.

Reply:
> user mapping for password authentication is useless, since clients always send the X-Trino-User header.


That's not correct assumption, see `--session-user`

> This happens on the server: [ ... ]
> Without changing all of the clients, there is no way for user mapping to work. We can safely assume that impersonating the password user was not intended behavior.

If user mapping does not work with password auth for some reason, users should not use the two together.

Reply:
> That's not correct assumption, see `--session-user`

`--session-user` does not change the situation. It allows users to impersonate a different user. However, existing clients, and new clients with `--session-user` when that flag is not specified, will still specify `X-Trino-User` as a side effect of password authentication, not because they actually want to impersonate the password user.

> If user mapping does not work with password auth for some reason, users should not use the two together.

I don't understand what you're saying here. It doesn't work, due to the behavior of all existing clients, as I explained above. Do you disagree with the above explanation of the current client and server behavior? Or do you disagree with the conclusion that this makes user mapping useless for password authentication?

The fundamental problem is that we effectively changed the model when we introduced user mapping, and it wasn't compatible with how existing clients do password authentication. This PR allows it to work, by working around the behavior of existing clients. (including clients that are outside of our control, or difficult to change, such as the commercial ODBC driver)

Reply:
I think @findepi has a good point here, using `--session-user` overrides the `X-Trino-User` header thus the user can impersonate itself but I would consider this as a workaround rather than the solution to the problem.

So what's the solution then? It looks like `X-Trino-User` header is required, or at least, always provided by clients. If I understand it correctly `X-Trino-User` header is actually required only if no authentication method is enabled as Trino has to know which user has sent the request. In any other case it's being used for impersonation purposes which makes it redundant or even obsolete in many cases such as this one.

Reply:
> It looks like X-Trino-User header is required, or at least, always provided by clients. 

iirc @dain or @electrum had some thoughts about this